File: web/src/next/engine/mock/tides.ts
/**
 * MOCK tide predictions (tides agent) for developing the interface only: one synthetic
 * harmonic station, "Mock Harbor", with five constituents and no nodal corrections.
 * Every number is illustrative. The shapes, units, error codes and limits are the real
 * engine's (EXPLORER_API "Tides", `skyfix_wasm::tides`).
 */

import { isoUtc } from '../../time.js';
import {
  TIDE_LABEL,
  type TideCurve,
  type TideDatum,
  type TideEvent,
  type TideExtremes,
  type TideNow,
  type TidesEngine,
  type TidesPackInfo,
  type TideStation,
  type TideStationNear,
} from '../types.js';

const DEG = Math.PI / 180;

/** Constituents of the synthetic tide: amplitude m, speed deg/h, phase deg (about MSL). */
const TERMS: readonly (readonly [number, number, number])[] = [
  [0.75, 28.9841042, 30], // M2
  [0.13, 30.0, 60], // S2
  [0.17, 28.4397295, 10], // N2
  [0.1, 15.0410686, 200], // K1
  [0.08, 13.9430356, 180], // O1
];

/** Datum offsets from MSL, metres. */
const DATUMS: Readonly<Record<TideDatum, number>> = {
  HAT: 1.2,
  MHHW: 0.95,
  MHW: 0.85,
  MTL: 0.0,
  MSL: 0.0,
  MLW: -0.85,
  MLLW: -0.92,
  LAT: -1.2,
  NAVD88: 0.05,
};

export const MOCK_TIDE_STATION: TideStation = {
  id: 'MOCK001',
  name: 'Mock Harbor (synthetic)',
  state: null,
  lat_deg: 39.9,
  lon_deg: -75.13,
  kind: 'harmonic',
  reference_id: null,
  reference_name: null,
  tide_type: 'semidiurnal',
  form_number: 0.2045,
  datums: ['HAT', 'MHHW', 'MHW', 'MTL', 'MSL', 'MLW', 'MLLW', 'LAT', 'NAVD88'],
  default_datum: 'MLLW',
  curve: 'harmonic',
  flags: [],
  notes: ['Synthetic station of the mock engine: every number is illustrative.'],
};

const FIRST_JD = 2415020.5; // 1900-01-01
const END_JD = 2488434.5; // 2101-01-01
const EPOCH = 2451545.0;

function fail(code: string, message: string): never {
  throw new Error(`${code}: ${message}`);
}

function height(jd: number): number {
  const tau = (jd - EPOCH) * 24;
  let h = 0;
  for (const [a, w, g] of TERMS) h += a * Math.cos((w * tau + g) * DEG);
  return h;
}

function rate(jd: number): number {
  const tau = (jd - EPOCH) * 24;
  let r = 0;
  for (const [a, w, g] of TERMS) r -= a * w * DEG * Math.sin((w * tau + g) * DEG);
  return r;
}

function checkJd(jd: number, what: string): void {
  if (!Number.isFinite(jd)) fail('bad_request', `${what} is not a number`);
  if (jd < FIRST_JD || jd >= END_JD) {
    fail('outside_range', `tide predictions are offered from 1900 to 2100; ${what} is ${isoUtc(jd)}`);
  }
}

function checkWindow(a: number, b: number): void {
  checkJd(a, 'jd_start');
  checkJd(b, 'jd_end');
  if (b < a) fail('bad_request', 'jd_end is before jd_start');
}

function resolveDatum(datum: string | undefined): TideDatum {
  const d = (datum ?? '').trim().toUpperCase();
  if (d === '') return 'MLLW';
  const name = (d === 'NAVD' ? 'NAVD88' : d) as TideDatum;
  if (!(name in DATUMS)) {
    fail('bad_request', `unknown datum ${JSON.stringify(datum)}; expected one of ${Object.keys(DATUMS).join(', ')}`);
  }
  return name;
}

function station(id: string): TideStation {
  if (id.trim().toUpperCase() !== MOCK_TIDE_STATION.id) {
    fail('unknown_station', `no tide station ${JSON.stringify(id.trim())} in the tides-us pack`);
  }
  return MOCK_TIDE_STATION;
}

function notes(d: TideDatum): string[] {
  return [`Heights above ${d} (mock engine: illustrative).`];
}

/** Extremes by the sign of the rate on a 6-minute grid, refined by bisection. */
function extremes(a: number, b: number, off: number): TideEvent[] {
  const step = 6 / 1440;
  const out: TideEvent[] = [];
  let t0 = a;
  let r0 = rate(t0);
  while (t0 < b) {
    const t1 = Math.min(t0 + step, b);
    const r1 = rate(t1);
    if ((r0 > 0 && r1 <= 0) || (r0 < 0 && r1 >= 0)) {
      let lo = t0;
      let hi = t1;
      for (let k = 0; k < 40; k += 1) {
        const mid = 0.5 * (lo + hi);
        if (rate(mid) > 0 === r0 > 0) lo = mid;
        else hi = mid;
      }
      const t = 0.5 * (lo + hi);
      out.push({ kind: r0 > 0 ? 'high' : 'low', jd_utc: t, utc: isoUtc(t), height_m: height(t) - off });
    }
    t0 = t1;
    r0 = r1;
  }
  return out;
}

export interface MockTidesOptions {
  /** Answer as if the tides-us pack were loaded (default true); false exercises `pack_not_loaded`. */
  loaded?: boolean;
}

/** The mock's tide engine: every method of `TidesEngine`. */
export class MockTides implements TidesEngine {
  private loaded: boolean;

  constructor(options: MockTidesOptions = {}) {
    this.loaded = options.loaded ?? true;
  }

  private need(): void {
    if (!this.loaded) {
      fail('pack_not_loaded', 'tide predictions need the tides-us pack (US stations, NOAA), which is not loaded');
    }
  }

  tideStationsNear(latDeg: number, lonDeg: number, n: number): TideStationNear[] {
    this.need();
    if (!(Number.isFinite(latDeg) && Math.abs(latDeg) <= 90)) fail('bad_request', `latitude ${latDeg} is outside -90..90`);
    if (!(Number.isFinite(lonDeg) && Math.abs(lonDeg) <= 360)) {
      fail('bad_request', `longitude ${lonDeg} is outside -360..360`);
    }
    void n; // one station: n (clamped to 1..100) cannot matter
    const s = MOCK_TIDE_STATION;
    const p1 = latDeg * DEG;
    const p2 = s.lat_deg * DEG;
    const dl = (s.lon_deg - lonDeg) * DEG;
    const h = Math.sin((p2 - p1) / 2) ** 2 + Math.cos(p1) * Math.cos(p2) * Math.sin(dl / 2) ** 2;
    const km = 2 * 6371.0088 * Math.asin(Math.min(1, Math.sqrt(h)));
    const y = Math.sin(dl) * Math.cos(p2);
    const x = Math.cos(p1) * Math.sin(p2) - Math.sin(p1) * Math.cos(p2) * Math.cos(dl);
    const bearing = ((Math.atan2(y, x) / DEG) % 360 + 360) % 360;
    return [{ ...s, distance_km: km, distance_nm: km / 1.852, bearing_deg: bearing }];
  }

  tideStation(stationId: string): TideStation {
    this.need();
    return station(stationId);
  }

  tidePredict(stationId: string, jdStart: number, jdEnd: number, stepMin: number, datum: TideDatum | '' = ''): TideCurve {
    this.need();
    const s = station(stationId);
    checkWindow(jdStart, jdEnd);
    if (!(Number.isFinite(stepMin) && stepMin >= 0.5 && stepMin <= 1440)) {
      fail('bad_request', 'step_min must be between 0.5 and 1440 minutes');
    }
    const d = resolveDatum(datum);
    const step = stepMin / 1440;
    const n = Math.floor((jdEnd - jdStart) / step + 1e-9) + 1;
    if (n > 20_000) fail('bad_request', `${n} samples requested; at most 20000 per call`);
    const jd = new Float64Array(n);
    const h = new Float64Array(n);
    for (let k = 0; k < n; k += 1) {
      jd[k] = jdStart + k * step;
      h[k] = height(jd[k]!) - DATUMS[d];
    }
    return {
      station: s,
      datum: d,
      method: 'harmonic',
      jd_start: jdStart,
      jd_end: jdEnd,
      step_min: stepMin,
      jd_utc: jd,
      height_m: h,
      label: TIDE_LABEL,
      notes: notes(d),
    };
  }

  tideExtremes(stationId: string, jdStart: number, jdEnd: number, datum: TideDatum | '' = ''): TideExtremes {
    this.need();
    const s = station(stationId);
    checkWindow(jdStart, jdEnd);
    if (jdEnd - jdStart > 400) fail('bad_request', 'at most 400 days of high and low water per call');
    const d = resolveDatum(datum);
    return {
      station: s,
      datum: d,
      method: 'harmonic',
      jd_start: jdStart,
      jd_end: jdEnd,
      extremes: extremes(jdStart, jdEnd, DATUMS[d]),
      label: TIDE_LABEL,
      notes: notes(d),
    };
  }

  tideNow(stationId: string, jdUtc: number, datum: TideDatum | '' = ''): TideNow {
    this.need();
    const s = station(stationId);
    checkJd(jdUtc, 'jd_utc');
    const d = resolveDatum(datum);
    const around = extremes(jdUtc - 1.5, jdUtc + 3, DATUMS[d]);
    const after = around.filter((e) => e.jd_utc > jdUtc);
    const before = around.filter((e) => e.jd_utc <= jdUtc);
    const r = rate(jdUtc);
    const next = after[0] ?? null;
    return {
      station: s,
      datum: d,
      method: 'harmonic',
      jd_utc: jdUtc,
      utc: isoUtc(jdUtc),
      height_m: height(jdUtc) - DATUMS[d],
      rate_m_per_h: r,
      state: r > 0 || (r === 0 && next?.kind === 'high') ? 'rising' : 'falling',
      previous: before.at(-1) ?? null,
      next,
      next_high: after.find((e) => e.kind === 'high') ?? null,
      next_low: after.find((e) => e.kind === 'low') ?? null,
      label: TIDE_LABEL,
      notes: notes(d),
    };
  }

  tidePackInfo(): TidesPackInfo | null {
    if (!this.loaded) return null;
    return {
      name: 'tides-us',
      version: 'mock',
      bytes: 0,
      provides: ['tides:us-noaa'],
      stations: 1,
      harmonic: 1,
      subordinate: 0,
    };
  }

  /** Accepts any bytes, as the mock's `loadPack` does. */
  loadTidesPack(bytes: Uint8Array): TidesPackInfo {
    void bytes;
    this.loaded = true;
    return this.tidePackInfo()!;
  }
}

File: crates/skyfix-cli/src/commands/explorer/tides.rs
//! Tides from the optional `tides-us` pack: `tide-stations`, `tide-station`,
//! `tide-predict`, `tide-extremes`, `tide-now` and `tide-pack`. OWNER: cli3 agent.
//!
//! Engine: `skyfix_tides` through the WASM adapter's `skyfix_wasm::tides::native`, with
//! the station database the pack installs (`--pack web/public/data/packs/tides-us`, as
//! the site loads its saved copy); wire format EXPLORER_API.md "Expansion programme —
//! tides"; definitions CONVENTIONS 13.11. Every call without the pack fails with the
//! engine's `pack_not_loaded: …` and a line saying how to load it. Predictions, not
//! observations: weather and surge are not included, and every report says so.

use anyhow::{Result, anyhow};
use skyfix_tides::TideDb;
use skyfix_tides::api::{
    TideCurve, TideEvent, TideExtremes, TideNow, TideStation, TideStationNear,
};
use skyfix_wasm::tides::native;

use super::args::{FormatArgs, PositionArgs, parse_instant};
use super::sun::WindowFlags;
use super::text;
use super::wire::{
    Align, Table, ZoneFlag, emit_json, push_field, push_note, shown_in, when_cells, when_headers,
};
use super::zone::ResolvedZone;
use crate::exit;
use crate::report;

/// The installed database, or the engine's own refusal with how to load the pack.
fn db() -> Option<std::sync::Arc<TideDb>> {
    native::installed()
}

fn with_hint<T>(r: std::result::Result<T, String>) -> Result<T> {
    r.map_err(|e| {
        if e.starts_with("pack_not_loaded") {
            anyhow!(
                "{e}\nLoad it for this run with --pack web/public/data/packs/tides-us (from the \
                 repository root), or --pack FILE for a copy of the pack"
            )
        } else {
            anyhow!(e)
        }
    })
}

/// `--datum`.
#[derive(clap::Args, Debug, Clone)]
pub struct DatumFlag {
    /// The datum heights are above: MLLW, MLW, MSL, MTL, MHW, MHHW, LAT, HAT or NAVD88.
    /// Default: the station's own (MLLW; MSL where NOAA publishes no datums).
    #[arg(long, value_name = "DATUM", default_value = "")]
    pub datum: String,
}

fn station_block(out: &mut String, s: &TideStation) {
    push_field(
        out,
        "Station",
        &format!(
            "{}{}, NOAA {}",
            s.name,
            s.state
                .as_deref()
                .map(|st| format!(", {st}"))
                .unwrap_or_default(),
            s.id
        ),
    );
    let pos = skyfix_core::types::LatLon {
        lat_deg: s.lat_deg,
        lon_deg: s.lon_deg,
    };
    push_field(
        out,
        "Place",
        &format!(
            "{} ({})",
            report::format_position(pos),
            report::format_position_decimal(pos)
        ),
    );
    let kind = match (s.kind, &s.reference_name, &s.reference_id) {
        ("subordinate", Some(n), Some(id)) => {
            format!("subordinate to {n} ({id}): times and heights by NOAA's offsets")
        }
        (k, _, _) => k.to_string(),
    };
    push_field(
        out,
        "Kind",
        &format!(
            "{kind}; {}{}",
            s.tide_type
                .map(|t| format!("{} tide", t.replace('_', " ")))
                .unwrap_or_else(|| "tide type unknown".to_string()),
            s.form_number
                .map(|f| format!(" (form number {f:.2})"))
                .unwrap_or_default()
        ),
    );
    push_field(
        out,
        "Datums",
        &format!("{} (default {})", s.datums.join(", "), s.default_datum),
    );
}

fn feet(m: f64) -> String {
    format!("{:.2}", m / 0.3048)
}

// ---------------------------------------------------------------------------
// tide-stations
// ---------------------------------------------------------------------------

#[derive(clap::Args, Debug)]
pub struct StationsNearArgs {
    #[command(flatten)]
    pub position: PositionArgs,
    /// How many stations, nearest first (1 to 100).
    #[arg(long, value_name = "N", default_value_t = 10)]
    pub count: u32,
    #[command(flatten)]
    pub format: FormatArgs,
}

pub fn run_stations_near(a: &StationsNearArgs) -> Result<u8> {
    let db = db();
    let r = with_hint(native::stations_near_in(
        db.as_deref(),
        a.position.lat,
        a.position.lon,
        a.count,
    ))?;
    if a.format.is_json() {
        emit_json(&r)?;
    } else {
        report::emit(&render_near(&r, a))?;
    }
    Ok(exit::OK)
}

fn render_near(r: &[TideStationNear], a: &StationsNearArgs) -> String {
    let p = skyfix_core::types::LatLon {
        lat_deg: a.position.lat,
        lon_deg: a.position.lon,
    };
    let mut out = format!(
        "TIDE STATIONS NEAR {} ({})\n\n",
        report::format_position(p),
        report::format_position_decimal(p)
    );
    let mut t = Table::new(&[
        ("id", Align::Left),
        ("station", Align::Left),
        ("state", Align::Left),
        ("NM", Align::Right),
        ("km", Align::Right),
        ("bearing", Align::Right),
        ("kind", Align::Left),
        ("curve", Align::Left),
    ]);
    for s in r {
        t.row(vec![
            s.station.id.clone(),
            s.station.name.clone(),
            s.station.state.clone().unwrap_or_default(),
            format!("{:.1}", s.distance_nm),
            format!("{:.1}", s.distance_km),
            text::course(s.bearing_deg.round()),
            s.station.kind.to_string(),
            s.station.curve.to_string(),
        ]);
    }
    out.push_str(&t.render("  "));
    out.push('\n');
    push_note(
        &mut out,
        "Distances on the sphere, bearings the initial great-circle bearing from the place, \
         true. curve: harmonic, a true tide curve; interpolated (a subordinate station), the \
         curve between its high and low waters, an estimate; none, no prediction.",
    );
    out
}

// ---------------------------------------------------------------------------
// tide-station
// ---------------------------------------------------------------------------

#[derive(clap::Args, Debug)]
pub struct StationArgs {
    /// The NOAA station id, e.g. 9414290.
    #[arg(value_name = "ID")]
    pub id: String,
    #[command(flatten)]
    pub format: FormatArgs,
}

pub fn run_station(a: &StationArgs) -> Result<u8> {
    let db = db();
    let s = with_hint(native::station_in(db.as_deref(), &a.id))?;
    if a.format.is_json() {
        emit_json(&s)?;
    } else {
        let mut out = String::from("TIDE STATION\n");
        station_block(&mut out, &s);
        push_field(&mut out, "Curve", s.curve);
        if !s.flags.is_empty() {
            push_field(&mut out, "Flags", &s.flags.join(", "));
        }
        if !s.notes.is_empty() {
            out.push('\n');
            for n in &s.notes {
                push_note(&mut out, n);
            }
        }
        report::emit(&out)?;
    }
    Ok(exit::OK)
}

/// The zone for a station: the nautical zone takes the station's longitude.
fn station_zone(zone: &ZoneFlag, id: &str) -> Result<(ResolvedZone, TideStation)> {
    let db = db();
    let s = with_hint(native::station_in(db.as_deref(), id))?;
    Ok((zone.resolve(s.lon_deg), s))
}

fn label_notes(out: &mut String, label: &str, notes: &[String]) {
    out.push('\n');
    push_note(out, &format!("{label}."));
    for n in notes {
        push_note(out, n);
    }
}

// ---------------------------------------------------------------------------
// tide-extremes
// ---------------------------------------------------------------------------

#[derive(clap::Args, Debug)]
pub struct ExtremesArgs {
    /// The NOAA station id.
    #[arg(value_name = "ID")]
    pub id: String,
    #[command(flatten)]
    pub window: WindowFlags,
    #[command(flatten)]
    pub datum: DatumFlag,
    #[command(flatten)]
    pub zone: ZoneFlag,
    #[command(flatten)]
    pub format: FormatArgs,
}

pub fn run_extremes(a: &ExtremesArgs) -> Result<u8> {
    let (zone, _) = station_zone(&a.zone, &a.id)?;
    let (s, e) = a.window.resolve(&zone)?;
    let db = db();
    let r = with_hint(native::extremes_in(
        db.as_deref(),
        &a.id,
        s,
        e,
        &a.datum.datum,
    ))?;
    if a.format.is_json() {
        emit_json(&r)?;
    } else {
        report::emit(&render_extremes(&r, &zone))?;
    }
    Ok(exit::OK)
}

fn event_table(events: &[TideEvent], zone: &ResolvedZone, datum: &str) -> String {
    let mut cols = when_headers(zone);
    cols.extend([
        ("tide", Align::Left),
        ("m", Align::Right),
        ("ft", Align::Right),
    ]);
    let mut t = Table::new(&cols);
    for ev in events {
        let mut row = when_cells(ev.jd_utc, zone);
        row.extend([
            ev.kind.to_string(),
            format!("{:.3}", ev.height_m),
            feet(ev.height_m),
        ]);
        t.row(row);
    }
    if t.is_empty() {
        return "  none in this window\n".to_string();
    }
    let mut s = t.render("  ");
    s.push_str(&format!("  heights above {datum}\n"));
    s
}

fn render_extremes(r: &TideExtremes, zone: &ResolvedZone) -> String {
    let mut out = String::from("HIGH AND LOW WATER\n");
    station_block(&mut out, &r.station);
    push_field(
        &mut out,
        "Window",
        &format!(
            "{} to {}{}; {}",
            text::utc(r.jd_start),
            text::utc(r.jd_end),
            shown_in(zone),
            if r.method == "subordinate_offsets" {
                "by the reference station's tides and NOAA's offsets"
            } else {
                "harmonic prediction"
            }
        ),
    );
    out.push('\n');
    out.push_str(&event_table(&r.extremes, zone, r.datum));
    label_notes(&mut out, r.label, &r.notes);
    out
}

// ---------------------------------------------------------------------------
// tide-predict
// ---------------------------------------------------------------------------

#[derive(clap::Args, Debug)]
pub struct PredictArgs {
    /// The NOAA station id.
    #[arg(value_name = "ID")]
    pub id: String,
    #[command(flatten)]
    pub window: WindowFlags,
    /// Minutes between heights, 0.5 to 1440.
    #[arg(long, value_name = "MIN", default_value_t = 10.0)]
    pub step: f64,
    #[command(flatten)]
    pub datum: DatumFlag,
    #[command(flatten)]
    pub zone: ZoneFlag,
    #[command(flatten)]
    pub format: FormatArgs,
}

pub fn run_predict(a: &PredictArgs) -> Result<u8> {
    let (zone, _) = station_zone(&a.zone, &a.id)?;
    let (s, e) = a.window.resolve(&zone)?;
    let db = db();
    let r = with_hint(native::predict_in(
        db.as_deref(),
        &a.id,
        s,
        e,
        a.step,
        &a.datum.datum,
    ))?;
    if a.format.is_json() {
        // The export's object; its jd_utc and height_m are Float64Arrays there, arrays
        // here, and TideCurve serialises to exactly those keys.
        emit_json(&r)?;
    } else {
        report::emit(&render_predict(&r, &zone))?;
    }
    Ok(exit::OK)
}

fn render_predict(r: &TideCurve, zone: &ResolvedZone) -> String {
    let mut out = String::from("TIDE CURVE\n");
    station_block(&mut out, &r.station);
    push_field(
        &mut out,
        "Window",
        &format!(
            "{} to {}{}, every {} min; {}",
            text::utc(r.jd_start),
            text::utc(r.jd_end),
            shown_in(zone),
            r.step_min,
            if r.method == "interpolated" {
                "interpolated between high and low water (NOAA's Table 3): an estimate"
            } else {
                "harmonic prediction"
            }
        ),
    );
    out.push('\n');
    let mut cols = when_headers(zone);
    cols.extend([("m", Align::Right), ("ft", Align::Right)]);
    let mut t = Table::new(&cols);
    for (jd, h) in r.jd_utc.iter().zip(&r.height_m) {
        let mut row = when_cells(*jd, zone);
        row.extend([format!("{h:.3}"), feet(*h)]);
        t.row(row);
    }
    if t.is_empty() {
        out.push_str("  no heights in this window\n");
    } else {
        out.push_str(&t.render("  "));
        out.push_str(&format!("  heights above {}\n", r.datum));
    }
    label_notes(&mut out, r.label, &r.notes);
    out
}

// ---------------------------------------------------------------------------
// tide-now
// ---------------------------------------------------------------------------

#[derive(clap::Args, Debug)]
pub struct NowArgs {
    /// The NOAA station id.
    #[arg(value_name = "ID")]
    pub id: String,
    /// The instant, RFC 3339 with a trailing Z.
    #[arg(long, value_name = "RFC3339", value_parser = parse_instant, allow_hyphen_values = true)]
    pub utc: f64,
    #[command(flatten)]
    pub datum: DatumFlag,
    #[command(flatten)]
    pub zone: ZoneFlag,
    #[command(flatten)]
    pub format: FormatArgs,
}

pub fn run_now(a: &NowArgs) -> Result<u8> {
    let (zone, _) = station_zone(&a.zone, &a.id)?;
    let db = db();
    let r = with_hint(native::now_in(db.as_deref(), &a.id, a.utc, &a.datum.datum))?;
    if a.format.is_json() {
        emit_json(&r)?;
    } else {
        report::emit(&render_now(&r, &zone))?;
    }
    Ok(exit::OK)
}

fn render_now(r: &TideNow, zone: &ResolvedZone) -> String {
    let mut out = String::from("THE TIDE NOW\n");
    station_block(&mut out, &r.station);
    push_field(&mut out, "Time", &text::utc(r.jd_utc));
    push_field(
        &mut out,
        "Tide",
        &format!(
            "{:.3} m ({} ft) above {}, {} at {:+.3} m an hour",
            r.height_m,
            feet(r.height_m),
            r.datum,
            r.state,
            r.rate_m_per_h
        ),
    );
    out.push('\n');
    let events: Vec<TideEvent> = [&r.previous, &r.next, &r.next_high, &r.next_low]
        .into_iter()
        .flatten()
        .cloned()
        .collect();
    let mut seen = std::collections::HashSet::new();
    let events: Vec<TideEvent> = events
        .into_iter()
        .filter(|e| seen.insert(e.utc.clone()))
        .collect();
    out.push_str(&event_table(&events, zone, r.datum));
    label_notes(&mut out, r.label, &r.notes);
    out
}

// ---------------------------------------------------------------------------
// tide-pack
// ---------------------------------------------------------------------------

#[derive(clap::Args, Debug)]
pub struct PackArgs {
    #[command(flatten)]
    pub format: FormatArgs,
}

pub fn run_pack(a: &PackArgs) -> Result<u8> {
    let info = native::pack_info();
    if a.format.is_json() {
        emit_json(&info)?;
        return Ok(exit::OK);
    }
    let out = match info {
        Some(i) => format!(
            "TIDES PACK\nPack       {} {}, {} bytes of data, provides {}\nStations   {} ({} \
             harmonic, {} subordinate)\n",
            i.name,
            i.version,
            i.bytes,
            i.provides.join(", "),
            i.stations,
            i.harmonic,
            i.subordinate
        ),
        None => "TIDES PACK\nPack       not loaded: give --pack web/public/data/packs/tides-us\n"
            .to_string(),
    };
    report::emit(&out)?;
    Ok(exit::OK)
}
